File: backend/app/infrastructure/database/mappers.py
from app.domain.entities.company import Company
from app.domain.entities.message import Message
from app.domain.entities.request import Request
from app.domain.entities.user import User
from app.infrastructure.database.models.company import CompanyModel
from app.infrastructure.database.models.message import MessageModel
from app.infrastructure.database.models.request import RequestModel
from app.infrastructure.database.models.user import UserModel


def company_to_domain(model: CompanyModel) -> Company:
    return Company(
        id=model.id,
        name=model.name,
        is_active=model.is_active,
        created_at=model.created_at,
        updated_at=model.updated_at,
    )


def company_to_model(entity: Company) -> CompanyModel:
    return CompanyModel(
        id=entity.id,
        name=entity.name,
        is_active=entity.is_active,
    )


def user_to_domain(model: UserModel) -> User:
    return User(
        id=model.id,
        company_id=model.company_id,
        name=model.name,
        email=model.email,
        password_hash=model.password_hash,
        role=model.role,
        is_active=model.is_active,
        created_at=model.created_at,
        updated_at=model.updated_at,
    )


def user_to_model(entity: User) -> UserModel:
    return UserModel(
        id=entity.id,
        company_id=entity.company_id,
        name=entity.name,
        email=entity.email,
        password_hash=entity.password_hash,
        role=entity.role,
        is_active=entity.is_active,
    )


def request_to_domain(model: RequestModel) -> Request:
    return Request(
        id=model.id,
        company_id=model.company_id,
        created_by=model.created_by,
        assigned_to=model.assigned_to,
        title=model.title,
        description=model.description,
        type=model.type,
        priority=model.priority,
        status=model.status,
        created_at=model.created_at,
        updated_at=model.updated_at,
        resolved_at=model.resolved_at,
    )


def request_to_model(entity: Request) -> RequestModel:
    return RequestModel(
        id=entity.id,
        company_id=entity.company_id,
        created_by=entity.created_by,
        assigned_to=entity.assigned_to,
        title=entity.title,
        description=entity.description,
        type=entity.type,
        priority=entity.priority,
        status=entity.status,
        resolved_at=entity.resolved_at,
    )


def message_to_domain(model: MessageModel) -> Message:
    return Message(
        id=model.id,
        request_id=model.request_id,
        author_id=model.author_id,
        content=model.content,
        created_at=model.created_at,
    )


def message_to_model(entity: Message) -> MessageModel:
    return MessageModel(
        id=entity.id,
        request_id=entity.request_id,
        author_id=entity.author_id,
        content=entity.content,
    )

File: frontend/features/requests/lib/mappers.ts
import type {
  BackendRequestPriority,
  BackendRequestStatus,
  BackendRequestType,
  RequestResponseBody,
} from "@/services/requests";
import type { MessageResponseBody } from "@/services/messages";
import type { Message, RequestCategory, RequestDetail, RequestPriority, RequestStatus } from "../types";
import { formatDate, formatDateTime } from "./format";

/**
 * The real `RequestResponse`/`MessageResponse` only expose `company_id`,
 * `created_by`, `assigned_to`, `author_id` as raw ids — no names. Resolving
 * them would require calling `GET /users`, which USER-role callers (the
 * only role that can reach `/requests`) get a 403 from. So company/
 * requester/assignee are intentionally NOT resolved to names here; the UI
 * derives "Tú" / "Sin asignar" style labels from these ids at render time
 * instead (see RequestActions/RequestMessages), never inventing a name.
 */

const typeToCategory: Record<BackendRequestType, RequestCategory> = {
  INCIDENT: "incident",
  QUESTION: "question",
  REQUEST: "request",
};

const categoryToType: Record<RequestCategory, BackendRequestType> = {
  incident: "INCIDENT",
  question: "QUESTION",
  request: "REQUEST",
};

const priorityMap: Record<BackendRequestPriority, RequestPriority> = {
  LOW: "low",
  MEDIUM: "medium",
  HIGH: "high",
  URGENT: "urgent",
};

const priorityToBackend: Record<RequestPriority, BackendRequestPriority> = {
  low: "LOW",
  medium: "MEDIUM",
  high: "HIGH",
  urgent: "URGENT",
};

const statusMap: Record<BackendRequestStatus, RequestStatus> = {
  PENDING: "pending",
  IN_PROGRESS: "in_progress",
  RESOLVED: "resolved",
  CANCELLED: "cancelled",
};

export function mapCategoryToBackend(category: RequestCategory): BackendRequestType {
  return categoryToType[category];
}

export function mapPriorityToBackend(priority: RequestPriority): BackendRequestPriority {
  return priorityToBackend[priority];
}

export function toRequestDetail(dto: RequestResponseBody): RequestDetail {
  return {
    id: dto.id,
    title: dto.title,
    description: dto.description,
    category: typeToCategory[dto.type],
    priority: priorityMap[dto.priority],
    status: statusMap[dto.status],
    createdAt: formatDate(dto.created_at),
    updatedAt: formatDate(dto.updated_at),
    resolvedAt: dto.resolved_at ? formatDate(dto.resolved_at) : null,
    companyId: dto.company_id,
    createdBy: dto.created_by,
    assignedTo: dto.assigned_to,
  };
}

export function toMessage(dto: MessageResponseBody): Message {
  return {
    id: dto.id,
    authorId: dto.author_id,
    content: dto.content,
    createdAt: formatDateTime(dto.created_at),
  };
}
